### ISSUE TITLE
Add ROUGE metrics

### ISSUE THREAD

**Issue by mattdangerw** [keras-team/keras-nlp#67]
Splitting this issue out from https://github.com/keras-team/keras-nlp/issues/38.

We should add a ROUGE metric as `keras_nlp.metrics.RougeL`, etc.

**Comment by mattdangerw** [comment]
From @abheesht17

> https://colab.research.google.com/drive/1xchVi4DsG_2gfi8FmO-g9GZS6-7CIFnz?usp=sharing and ROUGE-N: https://colab.research.google.com/drive/1jurZQeHH760TyOHkjjae-eqkjpC7qSTc?usp=sharing.
>
> I checked if the code was working in graph mode by adding @tf.function, so ideally, it should work during training as well. I'll add a training example tomorrow (probably the NMT example on the repo) to verify again.

**Comment by abheesht17** [comment]
Hey, @mattdangerw, @chenmoneygithub! If the notebooks look okay to you, should I open a PR?

**Comment by abheesht17** [comment]
@mattdangerw, could you please help me with the optimisation process?

These are the two notebooks:
ROUGE-N Notebook: https://colab.research.google.com/drive/1wiNN5KZaClr7qufAu3vzZtx6YECqsTrX?usp=sharing
ROUGE-L Notebook: https://colab.research.google.com/drive/1xchVi4DsG_2gfi8FmO-g9GZS6-7CIFnz?usp=sharing

The time taken for training per epochs blows up from seconds to hours with the current implementation. Any ideas on how I can optimise this?

**Comment by mattdangerw** [comment]
Will do! I'll be out tomorrow, so this will probably need to be next week, but I will try to carve some time this then.

**Comment by abheesht17** [comment]
Sure, take your time :)

**Comment by chenmoneygithub** [comment]
@abheesht17 I checked out your implementation, and the bottleneck is at your nested for-loop. 

For example, when you are extracting the n-gram, you are iterating over batch dim and then sequence dim to find candidate n-grams, then you use another for-loop to check if the n-gram is a duplication. The check can be done using a hashset. 

If you want to know exactly how the execution time is broken down, this tutorial could help: https://www.tensorflow.org/tensorboard/tensorboard_profiling_keras

Briefly you will need to have a dummy model trained on dummy data, and pass your ROUGE metric into `model.fit()`. Then you can check what is your bottleneck on the profile page.

**Comment by abheesht17** [comment]
Hey. @chenmoneygithub! Yeah, that is the major issue, which I've pointed out here: https://github.com/keras-team/keras-nlp/issues/67#issuecomment-1079714052 and here: https://github.com/keras-team/keras-nlp/pull/69#issue-1182398730.

But if I use a hashset, won't it break graph ops?

**Comment by chenmoneygithub** [comment]
We have to try it out, I do not know if hash set is allowed in graph execution. But there seems to be a tensorflow specific set: https://www.tensorflow.org/api_docs/python/tf/sets Maybe try this?

**Comment by mattdangerw** [comment]
Some thoughts...

 - TF Text actually has a rouge_l op implementation. https://www.tensorflow.org/text/api_docs/python/text/metrics/rouge_l Have you tried it? The longest common sequence is probably the hardest thing to recreate in a tf op, so this might unblock us.
 - For the rouge-n, it does seems like we might be able to play around with some ops in tensorflow to see if we can get what we need. Use `tf.fingerprint` and `tf.unique_with_counts ` as a way to count hash collisions of ngrams? Or `tf.sets.intersection` as Chen was pointing out? This definitely seems quite tricky! But I think we may not be totally blocked here.

If it ends up seeming like we really can't do a efficient op level solution, we could try for a slower python only approach and improve from there. But let's check out the above possibilities a bit first.

**Comment by abheesht17** [comment]
Sure. Thanks a ton, @mattdangerw. I'll try these things out.

**Comment by chenmoneygithub** [comment]
@abheesht17 Hi! I want to check in if there is anything we can help with ROUGE-N metric. Would you mind sharing what methods you have tried? thanks!

**Comment by abheesht17** [comment]
Hey, @chenmoneygithub! I haven't written any code yet, but I've been thinking. I'll come up with some code in a couple of days.

1.  For `tf.sets.intersection()`: This function expects a Sparse Tensor. Here, we can make use of [tf.sparse.from_dense()](https://www.tensorflow.org/api_docs/python/tf/sparse/from_dense) and [tf.sparse.to_dense()](https://www.tensorflow.org/api_docs/python/tf/sparse/to_dense). I'll try this out.
2. If you see this [notebook](https://colab.research.google.com/drive/1wiNN5KZaClr7qufAu3vzZtx6YECqsTrX?usp=sharing), currently, I am finding the n-grams and pushing them to a tensor. Is there a function in TensorFlow which automatically forms chunks of a given size (similar to [torch.chunk()](https://pytorch.org/docs/stable/generated/torch.chunk.html))? This will greatly reduce the time taken, I suppose, since the TensorFlow implementation will be more efficient. I tried looking for such functions in TensorFlow, but couldn't find any, except for [tf.image.extract_patches()](https://www.tensorflow.org/api_docs/python/tf/image/extract_patches). I am not sure if it works for tensors with rank = 1 or 2, and I am not sure if we should be using this function, in any case.
